soroban-env-common/src/env.rs

```diff
@@ -461,6 +467,23 @@ macro_rules! call_macro_with_all_host_functions {
                 {"0", fn verify_sig_ed25519(x:Object, k:Object, s:Object) -> RawVal}
             }
 
+            mod prng "p" {
+                /// Return a u32 from the persistent, stateful PRNG associated
+                /// with the calling context, finalizing the PRNG if `finalize`
```

Comment:
why do we need the `finalize` functionality? I thought we'd expose something similar to `/dev/random` ?

Reply:
See comment in prng.rs -- the point of finalizing is to help contracts defend against callers that keep poking them to try to leak bits of state from their PRNG.

Reply:
Oh I see. We can discuss this more on the mailing list when we get to that CAP too.

My intuition is that if a contract expects to make some secure decision based on PRNG we should never allow the non finalize behavior as it may be under attack.

Maybe a better version of this for v1 is to *not* expose this granular finalize and instead disallow PRNGs to be generated across contexts:
* `I` calls a method on `J`
    * `J`'s PRNG gets initialized the first time a random number is requested.
    * `J` calls `next_u32` as many times as it wants
* as soon as `J` returns, its PRNG gets finalized (note that it may be more complicated than this if we allow re-entrancy)
* if `I` tries to call `J` and `J` tries to generate a random number -> error

This still has the same vulnerability "ie `I` can poke `J` but requires an attacker to submit many transactions, which is not hard to do (ie `call lottery.pickLotteryWinner(); if !checkIWon() trap()` ) but a lot more expensive. We could make it that this "finalize" behavior is sticky for an entire ledger too but then maybe this is a way to DoS those contracts (``call lottery.pickLotteryWinner(); trap("LOL")` ).

There is actually a third option that might be better than all of this in v1: don't allow PRNGs to be used from a subcall (this may be also useful in general?). This would mitigate all problems above as attackers can't wrap calls with custom logic/conditions.
edit: oh actually there is one way. people could attack the contract using gas in some situations. If PRNG yields 2 outcomes `A` or `B`, and the attacker wants the cheaper outcome `A` it could provide enough gas to execute `A` but not enough to execute `B`. To mitigate this, we'd need a way to charge "constant gas" in some situations (which should not be that hard to do - but work).

Reply:
I put `finalize` in the API exposed from host to guest so that the SDK could experiment with this question without the host having to change its behaviour.

Reply:
I think only option 1 can be implemented in the SDK?

Reply:
I doubt the SDK can predict when finalize needs to be called. The SDK won't know if the user is going to call another function that will need random data. What am I missing?

Reply:
I think the way you could do a simplified version of 1 is by forcing a pattern with RAAI so that if a block of code uses a PRNG you can force call finalize, so that any other block would fail to generate a PRNG?

Reply:
I thought about this more, and while option 3 for example makes it better by orders of magnitude
* none of those options can resist biasing from an attacker (so an attacker can get 100x or 1000x higher chance of winning than good actors)
* those options don't really get better over time as we improve the PRNG seed.

I think the one pattern that would satisfy both of these properties is a select & reveal scheme:
* contract picks random values in one transaction and record the tx hash at the same time (so we need to have a way to get the hash of the current tx) - the contract should not leak any information about its choice, and because only contracts have access to their own state the "choice" cannot be attacked
* in a different transaction, the contract can act in a deterministic way based on choices made earlier.
     * For example, by seeding its local (deterministic) random generator with the random seed from step 1.
     * We'd need good pseudo-random library, but that should be easy to come by

Reply:
I don't think contracts can ever have state that's secret-for-long. As soon as a txset is confirmed its order is known and the consensus input to any PRNGs accessed in that block is known. The whole approach I'm aiming for here is that txs inside that txset might use the PRNG state decided _for_ the txset _in_ the txset so that it's too late to abuse that knowledge in the next txset (and too soon for adjacent txs inside the txset to know the inputs).

The commit/reveal schemes I've seen for building PRNGs on chain have to do with secret state held off-chain by 3rd party oracles. I don't think it can really work on-chain.

Reply:
Update: as predicted, people are "approximating" PRNG service by using various much-easier-to-guess constants such as pointer addresses. We should land something here ASAP.